File: components/api-http/src/main/resources/META-INF/dirigible/http/request.js
/*
 * Copyright (c) 2022 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2022 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */
 /**
 * HTTP API Request
 *
 */
 
const streams = require("io/streams");

let textData = null;

exports.isValid = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.isValid();
};

exports.getMethod = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getMethod();
};

exports.getRemoteUser = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getRemoteUser();
};

exports.getPathInfo = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getPathInfo();
};

exports.getPathTranslated = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getPathTranslated();
};

exports.getHeader = function (name) {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getHeader(name);
};

exports.isUserInRole = function (role) {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.isUserInRole(role);
};

exports.getAttribute = function (name) {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getAttribute(name);
};

exports.getAuthType = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getAuthType();
};

exports.getCookies = function () {
    let cookiesJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getCookies();
    return JSON.parse(cookiesJson);
};

exports.getAttributeNames = function () {
    let attrNamesJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getAttributeNames();
    return JSON.parse(attrNamesJson);
};

exports.getCharacterEncoding = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getCharacterEncoding();
};

exports.getContentLength = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getContentLength();
};

exports.getHeaders = function (name) {
    let headersJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getHeaders(name);
    return JSON.parse(headersJson);
};

exports.getContentType = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getContentType();
};

exports.getBytes = function () {
    let bytesJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getBytes();
    return JSON.parse(bytesJson);
};

var getText = exports.getText = function () {
    if (textData === null) {
        textData = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getText();
    }
    return textData;
};

exports.getJSON = function () {
    let text = getText();
    return JSON.parse(text);
};

exports.getParameter = function (name) {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getParameter(name);
};

exports.getParameters = function () {
    let paramsJson;
    paramsJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getParameters();
    return JSON.parse(paramsJson);
};

exports.getResourcePath = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getResourcePath();
};

exports.getHeaderNames = function () {
    let headerNamesJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getHeaderNames();
    return JSON.parse(headerNamesJson);
};

exports.getParameterNames = function () {
    let paramNamesJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getParameterNames();
    return JSON.parse(paramNamesJson);
};

exports.getParameterValues = function (name) {
    let paramValuesJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getParameterValues(name);
    return JSON.parse(paramValuesJson);
};

exports.getProtocol = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getProtocol();
};

exports.getScheme = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getScheme();
};

exports.getContextPath = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getContextPath();
};

exports.getServerName = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getServerName();
};

exports.getServerPort = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getServerPort();
};

var getQueryString = exports.getQueryString = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getQueryString();
};

/**
 * Returns the query string name value pairs as JS object map. When multiple query parameters with the same name are specified,
 * it will collect theirs values in an array in the order of declaration under that name in the map.
 */
exports.getQueryParametersMap = function () {
    let queryString = getQueryString();
    if (!queryString)
        return {};

    queryString = decodeURI(queryString);
    let queryStringSegments = queryString.split('&');

    let queryMap = {};
    queryStringSegments.forEach(function (seg) {
        seg = seg.replace('amp;', '');
        const kv = seg.split('=');
        const key = kv[0].trim();
        const value = kv[1] === undefined ? true : kv[1].trim();
        if (queryMap[key] !== undefined) {
            if (!Array.isArray(queryMap[key]))
                queryMap[key] = [queryMap[key]];
            else
                queryMap[key].push(value);
        } else {
            queryMap[key] = value;
        }
    }.bind(this));
    return queryMap;
};

exports.getRemoteAddress = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getRemoteAddress();
};

exports.getRemoteHost = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getRemoteHost();
};

exports.setAttribute = function (name, value) {
    org.eclipse.dirigible.components.api.http.HttpRequestFacade.setAttribute(name, value);
};

exports.removeAttribute = function (name) {
    org.eclipse.dirigible.components.api.http.HttpRequestFacade.removeAttribute(name);
};

exports.getLocale = function () {
    let localeJson = org.eclipse.dirigible.components.api.http.HttpRequestFacade.getLocale();
    return JSON.parse(localeJson);
};

exports.getRequestURI = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getRequestURI();
};

exports.isSecure = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.isSecure();
};

exports.getRequestURL = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getRequestURL();
};

exports.getServicePath = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getServicePath();
};

exports.getRemotePort = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getRemotePort();
};

exports.getLocalName = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getLocalName();
};

exports.getLocalAddress = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getLocalAddress();
};

exports.getLocalPort = function () {
    return org.eclipse.dirigible.components.api.http.HttpRequestFacade.getLocalPort();
};

exports.getInputStream = function () {
    return streams.createInputStream(org.eclipse.dirigible.components.api.http.HttpRequestFacade.getInputStream());
};

File: components/api/api-modules-python/src/main/resources/META-INF/dirigible/python-modules/dirigible/http/request.py
import json
import java

class Request:
    HttpRequestFacade = java.type('org.eclipse.dirigible.components.api.http.HttpRequestFacade')
    textData = None

    @staticmethod
    def isValid():
        return Request.HttpRequestFacade.isValid()

    @staticmethod
    def getMethod():
        return Request.HttpRequestFacade.getMethod()

    @staticmethod
    def getRemoteUser():
        return Request.HttpRequestFacade.getRemoteUser()

    @staticmethod
    def getPathInfo():
        return Request.HttpRequestFacade.getPathInfo()

    @staticmethod
    def getPathTranslated():
        return Request.HttpRequestFacade.getPathTranslated()

    @staticmethod
    def getHeader(name):
        return Request.HttpRequestFacade.getHeader(name)

    @staticmethod
    def isUserInRole(role):
        return Request.HttpRequestFacade.isUserInRole(role)

    @staticmethod
    def getAttribute(name):
        return Request.HttpRequestFacade.getAttribute(name)

    @staticmethod
    def getAuthType():
        return Request.HttpRequestFacade.getAuthType()

    @staticmethod
    def getCookies():
        cookiesJson = Request.HttpRequestFacade.getCookies()
        return json.loads(cookiesJson)

    @staticmethod
    def getAttributeNames():
        attrNamesJson = Request.HttpRequestFacade.getAttributeNames()
        return json.loads(attrNamesJson)

    @staticmethod
    def getCharacterEncoding():
        return Request.HttpRequestFacade.getCharacterEncoding()

    @staticmethod
    def getContentLength():
        return Request.HttpRequestFacade.getContentLength()

    @staticmethod
    def getHeaders(name):
        headersJson = Request.HttpRequestFacade.getHeaders(name)
        return json.loads(headersJson)

    @staticmethod
    def getContentType():
        return Request.HttpRequestFacade.getContentType()

    @staticmethod
    def getBytes():
        bytesJson = Request.HttpRequestFacade.getBytes()
        return json.loads(bytesJson)

    @staticmethod
    def getText():
        if Request.textData is None:
            Request.textData = Request.HttpRequestFacade.getText()
        return Request.textData

    @staticmethod
    def getJSON():
        try:
            text = Request.getText()
            return json.loads(text)
        except Exception as e:
            return None

    @staticmethod
    def getParameter(name):
        return Request.HttpRequestFacade.getParameter(name)

    @staticmethod
    def getParameters():
        paramsJson = Request.HttpRequestFacade.getParameters()
        return json.loads(paramsJson)

    @staticmethod
    def getResourcePath():
        return Request.HttpRequestFacade.getResourcePath()

    @staticmethod
    def getHeaderNames():
        headerNamesJson = Request.HttpRequestFacade.getHeaderNames()
        return json.loads(headerNamesJson)

    @staticmethod
    def getParameterNames():
        paramNamesJson = Request.HttpRequestFacade.getParameterNames()
        return json.loads(paramNamesJson)

    @staticmethod
    def getParameterValues(name):
        paramValuesJson = Request.HttpRequestFacade.getParameterValues(name)
        return json.loads(paramValuesJson)

    @staticmethod
    def getProtocol():
        return Request.HttpRequestFacade.getProtocol()

    @staticmethod
    def getScheme():
        return Request.HttpRequestFacade.getScheme()

    @staticmethod
    def getContextPath():
        return Request.HttpRequestFacade.getContextPath()

    @staticmethod
    def getServerName():
        return Request.HttpRequestFacade.getServerName()

    @staticmethod
    def getServerPort():
        return Request.HttpRequestFacade.getServerPort()

    @staticmethod
    def getQueryString():
        return Request.HttpRequestFacade.getQueryString()

    @staticmethod
    def getQueryParametersMap():
        queryString = Request.getQueryString()
        if not queryString:
            return {}

        queryString = decodeURI(queryString)
        queryStringSegments = queryString.split('&')

        queryMap = {}
        for seg in queryStringSegments:
            seg = seg.replace('amp;', '')
            kv = seg.split('=')
            key = kv[0].strip()
            value = kv[1] if len(kv) > 1 else True
            value = value.strip()
            if key in queryMap:
                if not isinstance(queryMap[key], list):
                    queryMap[key] = [queryMap[key]]
                queryMap[key].append(value)
            else:
                queryMap[key] = value

        return queryMap

    @staticmethod
    def getRemoteAddress():
        return Request.HttpRequestFacade.getRemoteAddress()

    @staticmethod
    def getRemoteHost():
        return Request.HttpRequestFacade.getRemoteHost()

    @staticmethod
    def setAttribute(name, value):
        Request.HttpRequestFacade.setAttribute(name, value)

    @staticmethod
    def removeAttribute(name):
        Request.HttpRequestFacade.removeAttribute(name)

    @staticmethod
    def getLocale():
        localeJson = Request.HttpRequestFacade.getLocale()
        return json.loads(localeJson)

    @staticmethod
    def getRequestURI():
        return Request.HttpRequestFacade.getRequestURI()

    @staticmethod
    def isSecure():
        return Request.HttpRequestFacade.isSecure()

    @staticmethod
    def getRequestURL():
        return Request.HttpRequestFacade.getRequestURL()

    @staticmethod
    def getServicePath():
        return Request.HttpRequestFacade.getServicePath()

    @staticmethod
    def getRemotePort():
        return Request.HttpRequestFacade.getRemotePort()

    @staticmethod
    def getLocalName():
        return Request.HttpRequestFacade.getLocalName()

    @staticmethod
    def getLocalAddress():
        return Request.HttpRequestFacade.getLocalAddress()

    @staticmethod
    def getLocalPort():
        return Request.HttpRequestFacade.getLocalPort()

    @staticmethod
    def getInputStream():
        raise Exception("getInputStream not implemented yet.")
        # TODO: implement "dirigible.streams" module
        # return streams.createInputStream(Request.HttpRequestFacade.getInputStream())
